pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -435,57 +422,102 @@ public Consumer getNextConsumer() {
-                    && currentConsumerRoundRobinIndex > resultingAvailableConsumerIndex)
-                            ? currentConsumerRoundRobinIndex : resultingAvailableConsumerIndex;
-
-            // if resultingAvailableConsumerIndex moved ahead of currentConsumerRoundRobinIndex: then we should
-            // check skipped consumer which had same priority as currentConsumerRoundRobinIndex consumer
-            boolean isLastConsumerBlocked = (currentConsumerRoundRobinIndex == (consumerList.size() - 1)
-                    && !isConsumerAvailable(consumerList.get(currentConsumerRoundRobinIndex)));
-            boolean shouldScanCurrentLevel = priorityLevel < 0
-                    /* means moved to next lower-priority-level */ || isLastConsumerBlocked;
-            if (shouldScanCurrentLevel && scanFromBeginningIfCurrentConsumerNotAvailable) {
-                for (int i = firstConsumerIndexOfCurrentPriorityLevel; i < currentConsumerRoundRobinIndex; i++) {
-                    Consumer nextConsumer = consumerList.get(i);
-                    if (isConsumerAvailable(nextConsumer)) {
-                        currentConsumerRoundRobinIndex = i + 1;
-                        return nextConsumer;
-                    }
+        // couldn't find available consumer
+        return null;
+    }
+
+    /**
+     * Finds index of first available consumer which has higher priority then given targetPriority
+     * @param targetPriority
+     * @return -1 if couldn't find any available consumer 
+     */
+    private int getConsumerFromHigherPriority(int targetPriority) {
+        for (int i = 0; i < currentConsumerRoundRobinIndex; i++) {
+            Consumer consumer = consumerList.get(i);
+            if (consumer.getPriorityLevel() < targetPriority) {
+                if (isConsumerAvailable(consumerList.get(i))) {
+                    return i;
                 }
-                // now, we have scanned from the beginning: flip the flag to avoid scan again
-                scanFromBeginningIfCurrentConsumerNotAvailable = false;
+            } else {
+                break;
             }
+        }
+        return -1;
+    }
 
-            Consumer nextConsumer = consumerList.get(resultingAvailableConsumerIndex);
-            if (isConsumerAvailable(nextConsumer)) {
-                currentConsumerRoundRobinIndex = resultingAvailableConsumerIndex + 1;
-                return nextConsumer;
+    /**
+     * Finds index of round-robin available consumer that present on same level as consumer on currentRoundRobinIndex if doesn't
+     * find consumer on same level then it finds first available consumer on lower priority level else returns index=-1
+     * if couldn't find any available consumer in the list
+     * 
+     * @param currentRoundRobinIndex
+     * @return
+     */
+    private int getNextConsumerFromSameOrLowerLevel(int currentRoundRobinIndex) {
+
+        int targetPriority = consumerList.get(currentRoundRobinIndex).getPriorityLevel();
```

Comment:
as we discussed: `getNextConsumerFromSameOrLowerLevel` checks consumer from current-level and saves the index of the next level in case we don't find consumer in the current-level and then does linear search in lower-level which is one loop so, as we discuss we will keep it in the same method.